src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs

```diff
@@ -0,0 +1,74 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.IO;
+using System.Reflection;
+
+namespace MongoDB.Driver.Core.NativeLibraryLoader
+{
+    internal abstract class RelativeLibraryLocatorBase : ILibraryLocator
+    {
+        // public methods
+        public string GetLibraryAbsolutePath(SupportedPlatform currentPlatform)
+        {
+            var relativePath = GetLibraryRelativePath(currentPlatform);
+            return GetAbsolutePath(relativePath);
+        }
+
+        public virtual Assembly GetLibraryBaseAssembly()
+        {
+            return typeof(RelativeLibraryLocatorBase).GetTypeInfo().Assembly;
+        }
+
+        public virtual string GetLibraryBasePath()
+        {
+            var assembly = GetLibraryBaseAssembly();
+            var codeBase = assembly.CodeBase;
+            var uri = new Uri(codeBase);
+            var absolutePath = uri.AbsolutePath;
+            return Path.GetDirectoryName(absolutePath);
+        }
+
+        public abstract string GetLibraryRelativePath(SupportedPlatform currentPlatform);
+
+        // private methods
+        private string GetAbsolutePath(string relativePath)
+        {
+            var libraryName = Path.GetFileName(relativePath);
+
+            var basePath = GetLibraryBasePath();
+            var absolutePath = Path.Combine(basePath, relativePath);
+
+            var currentFolderAbsolutePath = Path.GetFullPath(libraryName);
+
+            var locationsToCheck = new[]
+            {
+                currentFolderAbsolutePath,    // look at the current folder
+                absolutePath
+            };
+
+            foreach (var location in locationsToCheck)
+            {
+                if (File.Exists(location))
+                {
+                    return location;
+                }
+            }
+            
+            throw new FileNotFoundException($"Could not find library {libraryName}. Checked {string.Join(";", locationsToCheck)}.");
+        }
```

Comment:
I would refactor this method to something like:

    private string GetAbsolutePath(string relativePath)
    {
        var libraryBasePath = GetLibraryBasePath();
        var libraryName = Path.GetFileName(relativePath);

        var absolutePathsToCheck = new[]
        {
            Path.GetFullPath(libraryName), // look in the current folder
            Path.Combine(libraryBasePath, relativePath)
        };

        foreach (var absolutePath in absolutePathsToCheck)
        {
            if (File.Exists(absolutePath))
            {
                return absolutePath;
            }
        }
        
        throw new FileNotFoundException($"Could not find library {libraryName}. Checked {string.Join(";", absolutePathsToCheck)}.");
    }

Reply:
BUT... I still don't understand why we are looking in the current directory. The current directory should be treated as a random location since it is out of our control.

Reply:
BUT... I still don't understand why we are looking in the current directory. The current directory should be treated as a random location since it is out of our control.

Reply:
Discussed offline. Changed the current folder on the current assembly fodler

Reply:
** Per our offline discussion, @rstam is correct in that the current directory does not always equal the current assembly directory: The two locations we should search are relative to the current assembly location:
$ASSEMBLY_LOCATION\..\..\native\x64\windows\snappy64.dll 
$ASSEMBLY_LOCATION\snappy64.dll

So we just need to make this one final minor change :)